api/explorer/api/ingest.py

```diff
@@ -175,28 +143,49 @@ def ingest_nix_graph(nix_graph: model.NixGraph, g: GraphTraversalSource) -> None
             the Nix graph
     """
 
-    # Note: As an optimization we split logic for writing the graph into two
-    # functions. The first one ensures that every root node of the graph exists
-    def write_root_node_to_graph(root_package: graph.Package) -> None:
-        graph.insert_unique_vertex(root_package, g)
-
-    # Then, this second function handles writing the next level of the graph as well
-    # as the edge connecting the levels, assuming that the current level already exists
-    # in the Gremlin graph.
-    def write_layer_to_graph(
-        current_package: graph.Package,
-        upstream_package: graph.Package,
-        edge: graph.Edge,
-    ) -> None:
-        graph.insert_unique_vertex(upstream_package, g)
+    packages, edges = split_nix_graph(nix_graph)
+
+    # Write nodes to the Gremlin graph
+    click.echo("Writing nodes to the Gremlin graph...")
+
+    def get_local_client() -> GraphTraversalSource:
+        thread_local = threading.local()
+        g = getattr(thread_local, "g", None)
+        if g is None:
+            thread_local.g = g_factory()
+        return thread_local.g
+
+    def write_package(p: graph.Package):
+        # print(p)
+        g = get_local_client()
+        graph.insert_unique_vertex(p, g)
+
+    def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
+        current_package, edge, dependency_package = input
+        g = get_local_client()
         graph.insert_unique_directed_edge(
             edge,
             from_vertex=current_package,
-            to_vertex=upstream_package,
+            to_vertex=dependency_package,
             g=g,
         )
 
-    traverse(nix_graph, write_root_node_to_graph, write_layer_to_graph)
+    with ThreadPoolExecutor() as ex:
+        click.echo("Writing packages")
+        with click.progressbar(length=len(packages)) as bar:
+            futures = [ex.submit(write_package, p) for p in packages]
+            for f in futures:
+                f.add_done_callback(lambda _x: bar.update(1))
+            wait(futures)
+            # FIXME: Check futures for exceptions
```

Comment:
I think it's really important to at least assert that it all worked (but properly reporting these exceptions is probably not a priority).

Reply:
Agreed. I think in the case of vertices this implementation relies on all vertices successfully being written since `write_edge` will not create the neighboring vertices if they do not already exist. Perhaps it would make sense to allow exceptions to bubble up here, e.g. by re-raising the first encountered in `futures`, then swallow / log them when inserting edges below. Relevant docs for working with `Future` objects can be found here: https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.exception

Reply:
Thanks for the review!
I'm not sure if I understand your suggestions correctly :thinking: I've pushed the following changes: 
- for writing vertices:
  - Exceptions encountered while writing the vertices are collected in the exceptions list.
  - After all tasks have completed, if any exceptions are present in the exceptions list, a single exception is raised using the first exception encountered as the cause (from exceptions[0]).
- for writing edges:
  - use a logger to log exceptions as they occur
  
Please let me know if it looks good to you! @GuillaumeDesforges @dorranh